### PR TITLE
Refactor/fly 2245

### PR DESCRIPTION
**Pull request by Hakob23**
## What

Implement a two-step LP registration flow.
Added request lifecycle functions `approveRegistration`, `rejectRegistration`, and `withdrawRegisterRequest`, plus `getRegistrationState`  getter function, tests and fixture updates.

## Why

This change prevents immediate activation on `register` and guarantees collateral is returned when a request is rejected or withdrawn.

## Type of Change

- [ ] Bug fix (non-breaking change)
- [x] New feature (non-breaking change)
- [ ] Breaking change
- [ ] Refactor (no intended behavior change)
- [x] Security hardening
- [ ] Tests only
- [ ] Build/CI/Tooling
- [ ] Documentation

## Affected Areas

- [ ] PegIn flow (`registerPegIn`, `callForUser`, quote validation)
- [ ] PegOut flow (`depositPegout`, `refundPegOut`, `refundUserPegOut`)
- [x] Liquidity Provider lifecycle (register/update/status/resign/collateral/withdraw)
- [ ] Pause/operational controls
- [ ] Quote hashing/signature logic
- [ ] Bitcoin tx / PMT / merkle validation
- [ ] Deployment or upgrade scripts / Makefile targets
- [ ] CI workflows (`ci`, `fuzz`, `slither`, `codeql`)
- [ ] Docs

## Risk & Security Review

- [x] Access control and authorization paths reviewed
- [x] Reentrancy and external call paths reviewed
- [x] Storage layout / upgrade safety reviewed (if upgradeable code changed)
- [ ] Time/block/confirmation assumptions reviewed
- [ ] BTC tx output/index assumptions reviewed (if pegout validation touched)
- [x] No secrets or sensitive values added

Additional notes on test coverage, edge cases, and any tests intentionally skipped:

- Ran targeted and full Forge suites while implementing this change:
  - `forge test --match-path 'test/discovery/*.t.sol'`
  - `forge test --match-path 'test/integration/*.t.sol'`
  - `forge test --match-path test/Pause.t.sol`
  - `forge test --match-path test/Benchmark.t.sol`
  - `forge test` (full suite)
- Added/updated tests for:
  - pending registration not listed/operational until approved
  - owner approve path
  - owner reject path with collateral refund
  - LP withdraw request path with collateral refund
  - unauthorized approval rejection
  - re-registration behavior and fixture compatibility

## Deployment & Ops Impact

- [ ] No deployment impact
- [ ] Requires deployment
- [x] Requires upgrade
- [ ] Env/config changes required (`.env`, network RPCs, verification, etc.)
- [ ] Runbook/update instructions added to PR description

## Related Issues

- Jira: [FLY-2245](https://rsklabs.atlassian.net/browse/FLY-2245)

Key local outputs:

- Full `forge test` suite passed after updating registration-dependent tests and fixtures.

### CODE REVIEW

File: src/FlyoverDiscovery.sol [hunk 1]
```diff
@@ -29,6 +37,8 @@ contract FlyoverDiscovery is
 
     mapping(uint => Flyover.LiquidityProvider) private _liquidityProviders;
     mapping(address => uint) private _providerIdByAddress;
+    mapping(address => PendingRegistration) private _pendingRegistrations;
```
Comment (Luisfc68) [inline src/FlyoverDiscovery.sol]: This will be deployed in already existing contracts (since the 2.5.0 is already deployed) so we need to respect the storage layout from now on. This mappings should be moved to below the already existing variables

File: src/FlyoverDiscovery.sol [hunk 2]
```diff
@@ -77,19 +87,65 @@ contract FlyoverDiscovery is
...
-        _addCollateral(providerType, msg.sender, msg.value);
         return providerId;
     }
 
+    /// @inheritdoc IFlyoverDiscovery
+    function withdrawRegisterRequest() external whenNotPaused {
+        if (_registrationStates[msg.sender] != RegistrationState.Pending) {
+            revert RegistrationNotPending(msg.sender);
+        }
+        uint256 providerId = _providerIdByAddress[msg.sender];
```
Comment (Luisfc68) [inline src/FlyoverDiscovery.sol:109]: which is the rationale to give the provider id when the registration request is created? isn't it more intuitive (and simple) to give it only if the registration is approved?

File: src/FlyoverDiscovery.sol [hunk 3]
```diff
@@ -77,19 +87,65 @@ contract FlyoverDiscovery is
...
+        _registrationStates[providerAddress] = RegistrationState.Approved;
+        delete _pendingRegistrations[providerAddress];
+
+        emit RegistrationApproved(providerId, providerAddress, pending.collateralAmount);
+        _addCollateral(pending.providerType, providerAddress, pending.collateralAmount);
+    }
+
+    /// @inheritdoc IFlyoverDiscovery
+    function rejectRegistration(address providerAddress) external whenNotPaused {
+        if (msg.sender != defaultAdmin()) revert NotAuthorized(msg.sender);
```
Comment (Luisfc68) [inline src/FlyoverDiscovery.sol]: the admin check and pending state check repeats in all the registration management functions, maybe we should add a modifier, do you see any downsides?

File: src/FlyoverDiscovery.sol [hunk 4]
```diff
@@ -245,7 +326,9 @@ contract FlyoverDiscovery is
     /// @return The liquidity provider record, reverts if not found
     function _getProvider(address providerAddress) private view returns (Flyover.LiquidityProvider memory) {
         uint providerId = _providerIdByAddress[providerAddress];
-        if (providerId == 0) revert Flyover.ProviderNotRegistered(providerAddress);
+        if (providerId == 0 || _registrationStates[providerAddress] != RegistrationState.Approved) {
```
Comment (Luisfc68) [inline src/FlyoverDiscovery.sol:333]: For example, I think this change wouldn't be necessary if we granted the id only after the approval

**Review by Luisfc68 — CHANGES_REQUESTED** [review]

File: src/FlyoverDiscovery.sol (hunk 4 above)
Comment (Hakob23) [reply src/FlyoverDiscovery.sol:333]: Still, I think it is a more consistent way for tracking for the reasons mentioned above, and this check is not a big deal in gas wise but I am open for the other option as well

File: src/FlyoverDiscovery.sol (hunk 3 above)
Comment (Hakob23) [reply src/FlyoverDiscovery.sol]: I agree, but I have added a single private _check function instead of 1 modifier, because the 2 conditions are logically separate so it would make more sense to have 2 separate modifiers but in that case we would have a longer code(which contradicts to the initial idea of making the code more concise)

File: src/FlyoverDiscovery.sol (hunk 2 above)
Comment (Hakob23) [reply src/FlyoverDiscovery.sol:109]: Mainly so events and indexing stay consistent and approval stays a single write to _liquidityProviders without a new id assignment

File: src/FlyoverDiscovery.sol (hunk 3 above)
Comment (Hakob23) [reply src/FlyoverDiscovery.sol]: https://github.com/rsksmart/liquidity-bridge-contract/pull/449/changes/229fd9868dfad5a1ba5a3e87be3314366a641dbb

File: src/FlyoverDiscovery.sol (hunk 1 above)
Comment (Hakob23) [reply src/FlyoverDiscovery.sol]: https://github.com/rsksmart/liquidity-bridge-contract/pull/449/changes/5e10bdb21a7ae5dc465b53abb419272b6b612fc3

**Review by Luisfc68 — APPROVED** [review]
